engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,567 @@ package ingestion
+	if s.nodeVersion == nil {
+		return nil
+	}
 
-	s.state = StopControlPaused
+	if s.versionBeacon != nil && s.versionBeacon.SealHeight >= height {
+		// we already processed this or a higher version beacon
+		return nil
+	}
 
-	s.log.Warn().Msgf(
-		"Pausing execution as finalization reached "+
-			"the requested stop height %d",
-		s.stopHeight,
-	)
+	vb, err := s.versionBeacons.Highest(height)
+	if err != nil {
+		return fmt.Errorf("failed to get highest "+
+			"version beacon for stop control: %w", err)
+	}
 
-}
+	if vb == nil {
+		// no version beacon found
+		// this is unexpected as there should always be at least the
+		// starting version beacon, but not fatal.
+		// It can happen if the node starts before bootstrap is finished.
+		s.log.Info().
+			Uint64("height", height).
+			Msg("No version beacon found for stop control")
+		return nil
+	}
 
-// executingBlockHeight should be called while execution of height starts,
-// used for internal tracking of the minimum possible value of stopHeight
-func (s *StopControl) executingBlockHeight(height uint64) {
-	// TODO: should we lock here?
+	s.log.Info().
+		Uint64("vb_seal_height", vb.SealHeight).
+		Uint64("vb_sequence", vb.Sequence).
+		Msg("New version beacon found")
 
-	if s.state == StopControlPaused {
-		return
+	// this is now the last handled version beacon
+	s.versionBeacon = vb
+
+	// this is a new version beacon check what boundary it sets
+	stopHeight, shouldBeSet, err := s.getVersionBeaconStopHeight(vb)
+	if err != nil {
+		return err
+	}
+
+	set := s.stopBoundary != nil
+
+	if !set && !shouldBeSet {
+		// all good, no stop boundary set
+		return nil
+	}
+
+	var newStop *stopBoundary
+	if shouldBeSet {
+		newStop = &stopBoundary{
+			StopParameters: StopParameters{
+				StopHeight:  stopHeight,
+				ShouldCrash: s.crashOnVersionBoundaryReached,
+			},
+			fromVersionBeacon: true,
+		}
+	}
+
+	err = s.unsafeSetStopParameters(newStop, true)
+	if err != nil {
+		s.log.Warn().
+			Err(err).
+			Msg("Cannot change stop boundary when detecting new version beacon")
 	}
 
-	// updating the highest executing height, which will be used to reject setting
-	// stopHeight that is too low.
-	if height > s.highestExecutingHeight {
-		s.highestExecutingHeight = height
+	return nil
+}
+
+// getVersionBeaconStopHeight returns the stop height that should be set
+// based on the version beacon
+// error is not expected during normal operation since the version beacon
+// should have been validated when indexing
+func (s *StopControl) getVersionBeaconStopHeight(
+	vb *flow.SealedVersionBeacon,
+) (
+	uint64,
+	bool,
+	error,
+) {
+	// version boundaries are sorted by version
+	for _, boundary := range vb.VersionBoundaries {
+		ver, err := boundary.Semver()
+		if err != nil || ver == nil {
+			// this should never happen as we already validated the version beacon
+			// when indexing it
+			return 0, false, fmt.Errorf("failed to parse semver: %w", err)
+		}
+
+		if s.nodeVersion.LessThan(*ver) {
```

Comment:
I feel that using this conditional to decide which boundary to stop at next is nonideal. 

The desired behaviour, from my perspective, is: We alway stop at the closest version boundary whose `StopHeight` is greater than our local finalized height. The `StopControl` doesn't even need to care about versions at all, it is only responsible for stopping at a sequence of heights. 

Deciding the next stopping point based on semver comparison is _probably equivalent in practice_ to that desired behaviour, but we need to make additional assumptions about the version beacon, which are inconsistent with how it is defined so far.

For example, the VersionBeacon documentation says:
> While heights are strictly increasing, versions must be **equal or greater** when compared using semver semantics.

That means, it is valid for two consecutive version boundaries to be equivalent from the perspective of semver. If we were processing such a version beacon using the current logic, we would miss the second version boundary!

Reply:
I don't think there is an inconsistency here.

Let me try to illustrate why with an example:

Node version is V99 and latest finalised and executed block is 1 a version beacon arrives with the following boundaries (H - height, V - version):
- H0, V0
- H100, V100
- H200, V100 <- same version
- H300, V101

Even though the the versions are equal we have to stop at H100 not H200 so if we are forward iterating over the boundaries and comparing version with "less then" this is ok. 

When we hit H100 the node will stop, and should be restarted with V100 at which point the next boundary is at H200 but we don't need to stop there, since we are already at V100 so we will stop at H300 for the switch to V101


I disagree on this statement:
> We alway stop at the closest version boundary whose StopHeight is greater than our local finalized height. The StopControl doesn't even need to care about versions at all, it is only responsible for stopping at a sequence of heights.

My reasoning is that while I'm not 100% sure how we will define the versioning in the future, I can imagine that we might want to chose not to stop if only the patch (or minor) version has changed, since the node would still be able to participate in the network. IMO If a version change does not necessarily **require** that  all nodes switch to it at the same block, it is detrimental to stop  all the nodes at the same height, but rather do a rolling upgrade.

Reply:
> IMO If a version change does not necessarily require that all nodes switch to it at the same block, it is detrimental to stop all the nodes at the same height, but rather do a rolling upgrade.

I agree, but in this case why would we use the version beacon at all? The version beacon is useful when the exact version and block height are important to get right. If we want to do a rolling upgrade of a backward-compatible patch version, I don't think we need the version beacon system (or StopControl).

You're right about the edge case I proposed, though. I assumed semver would interpret (for example) `1.2.3-something` and `1.2.3-somethingelse` as the same version, but it does define an ordering on these. (The ordering would be essentially random given how we currently tag these kinds of releases, but that's a separate problem.)

Overall I still have a slight preference for using only block heights to decide when our next stop is, because I feel it is easier to reason about the behaviour and is more tolerant of version beacon inputs. But I'm happy to continue with how it is implemented here and agree it is correct. 👍 

I still feel we need to be able to enforce that an initial version is set ([this comment](https://github.com/onflow/flow-go/pull/4342#discussion_r1209618244)). I don't think it is safe to silently disable StopControl when we can't read our current version; I feel the StopControl should have to be explicitly turned off. Otherwise we are at risk of execution forks, if only a portion of ENs respect the version beacon output. We could address this in a follow-up PR, though.

Reply:
> I still feel we need to be able to enforce that an initial version is set

I agree, lets continue this on that comment.